### PR TITLE
Task2 update - Stack with generics

### PR DESCRIPTION
**Pull request by silvaby**
Hello!
This is updated version of Stack that uses generics.
This is a continuation of the PR #7 .
I would like to merge these branches.
Please leave your feedback.

**Task:**
Create a new class `Stack`. 
It should have the following interface:
* `func push(element: Int)` // Adds element on the top of the stack.
* `func pop() -> Int?` // Removes and returns the top element of the stack. returns nil when the invoking stack is empty.
* `func top() -> Int?` // Returns the top element of the stack, but does not remove it.

Create a new class `StackStatistics` that is a simple stack as described above but it also should support the following operation:
`func minimumElement() -> Int?` // Returns the first minimum element in the stack. Important: the function should work in O(1).

 Use inheritance. Use generics.

### CODE REVIEW

**Comment by silvaby** [comment]
I think that updates of task (pure Stack implementation) PR #3 is PR #9 (with generics).

File: Task2.playground/Contents.swift [hunk 1]
```diff
@@ -10,35 +10,63 @@
...
+    /// empty; otherwise, `nil`.
+    ///
+    /// - Complexity: O(1)
+    func pop() -> Element? {
         if items.isEmpty {
             return nil
         } else {
             return items.removeLast()
         }
     }
```
Comment (AndreyKovalevskiy) [inline Task2.playground/Contents.swift]: Please, read the documentation for Array(removing elements) and try to make this method easier. Here and below.

Comment (silvaby) [reply Task2.playground/Contents.swift]: Hi.
What does it mean "easier"? Tell me more information, please.

Comment (AndreyKovalevskiy) [reply Task2.playground/Contents.swift]: Instead several lines of code in this method you can use one of the methods of Array. Please read Array documentation.

File: Task2.playground/Contents.swift [hunk 2]
```diff
@@ -10,35 +10,63 @@
...
 
-class StackStatistics: Stack {
-    private var minimumElements = [Int]()
+class StackStatistics<Element: Comparable>: Stack<Element> {
+    private var minimumElements = [Element]()
 
-    override func push(element: Int) {
+    override func push(element: Element) {
         super.push(element: element)
         if minimumElements.isEmpty || element < minimumElements.last! {
```
Comment (TatsianaGouteeva) [inline Task2.playground/Contents.swift]: nit:

As I said earlier in https://github.com/silvaby/Hometasks/pull/7#discussion_r385636781 I don't like such implementation.

File: Task2.playground/Contents.swift [hunk 3]
```diff
@@ -47,34 +75,29 @@ class StackStatistics: Stack {
         }
     }
 
-    override func pop() -> Int? {
+    override func pop() -> Element? {
         if minimumElements.last != nil {
             minimumElements.removeLast()
         }
```
Comment (TatsianaGouteeva) [inline Task2.playground/Contents.swift]: Agree with @AndreyKovalevskiy .
Please, read the documentation for Array(removing elements) and try to make this method easier.

File: Task2.playground/Contents.swift (hunk 1 above)
Comment (silvaby) [reply Task2.playground/Contents.swift]: Done.

File: Task2.playground/Contents.swift (hunk 2 above)
Comment (silvaby) [reply Task2.playground/Contents.swift]: Done.

File: Task2.playground/Contents.swift (hunk 3 above)
Comment (silvaby) [reply Task2.playground/Contents.swift]: Done.

File: Task2.playground/Contents.swift [hunk 4]
```diff
@@ -10,71 +10,88 @@
...
+ */
 
-    // Adds element on the top of the stack.
-    func push(element: Int) {
+class Stack<Element> {
+    private var items = [Element]()
+
+    /// Adds the element on the top of the stack.
+    ///
+    /// Use this method to add a single element on the top of the stack.
```
Comment (dersim-davaod) [inline Task2.playground/Contents.swift]: Do you think it's useful to have almost the same line as above?

File: Task2.playground/Contents.swift [hunk 5]
```diff
@@ -10,71 +10,88 @@
...
-    private var items = [Int]()
+ Use generics.
+ */
 
-    // Adds element on the top of the stack.
-    func push(element: Int) {
+class Stack<Element> {
+    private var items = [Element]()
+
+    /// Adds the element on the top of the stack.
```
Comment (dersim-davaod) [inline Task2.playground/Contents.swift]: What does it mean "on the top of the stack"?

File: Task2.playground/Contents.swift [hunk 6]
```diff
@@ -10,71 +10,88 @@
...
+    ///
+    /// Use this method to add a single element on the top of the stack.
+    ///
+    ///     var someStack = [1, 2, 3]
```
Comment (dersim-davaod) [inline Task2.playground/Contents.swift]: Can you create stack in a such way?
I think your example can be confusing.

File: Task2.playground/Contents.swift [hunk 7]
```diff
@@ -10,71 +10,88 @@
...
+    ///     someStack.push(element: 10)
+    ///     print(someStack)
+    ///     // Prints "[1, 2, 3, 10]"
+    ///
+    /// - Parameter element: The element to add on the top of the stack.
```
Comment (dersim-davaod) [inline Task2.playground/Contents.swift]: nit:
You already described that the element will be added on the top.
```suggestion
    /// - Parameter element: The element to add to the stack.
```

File: Task2.playground/Contents.swift [hunk 8]
```diff
@@ -10,71 +10,88 @@
...
-        }
+    override func pop() -> Element? {
+        minimumElements.popLast()
         return super.pop()
     }
 
-    func minimumElement() -> Int? {
+    /// Returns the first minimum element in the stack.
+    ///
+    /// If the stack is empty, the method returns `nil`.
```
Comment (dersim-davaod) [inline Task2.playground/Contents.swift]: You already have this information in `Returns` section, don't you?

File: Task2.playground/Contents.swift [hunk 9]
```diff
@@ -10,71 +10,88 @@
...
-    private var minimumElements = [Int]()
+class StackStatistics<Element: Comparable>: Stack<Element> {
+    private var minimumElements = [Element]()
 
-    override func push(element: Int) {
+    override func push(element: Element) {
         super.push(element: element)
-        if minimumElements.isEmpty || element < minimumElements.last! {
-            minimumElements.append(element)
+        if let minimumElement = minimumElements.last {
```
Comment (dersim-davaod) [inline Task2.playground/Contents.swift]: I don't like this code, there's too much duplications. Try to simplify it. Smth like:

```swift
    lesserElement = minimumElement.last ?? element
    minimumElements.append(lesserElement)
```

**Pull request merged**